### PR TITLE
Initial commit of NATS key-value replicated in-memory cache

### PR DESCRIPTION
**Pull request by autodidaddict**

### CODE REVIEW

**Review by brooksmtownsend — COMMENTED** [review]
Overall LGTM, for some reason I get a linker error when compiling to x86_64-apple-darwin. No issues with using `cross` to build for linux, so more investigation to go there.

Only request, I think we should rename the provider from `nats-keyvalue` to something like `nats-cache` or `nats-replicated-cache` to avoid confusion, thoughts?

**Comment by brooksmtownsend** [comment]
`cargo build` and `cargo build --release` fails for me on x86_64-apple-darwin with angry clang:
```
  = note: Undefined symbols for architecture x86_64:
            "___guest_error", referenced from:
                ___guest_call in libwapc_guest-3f6684dfb69c655b.rlib(wapc_guest-3f6684dfb69c655b.wapc_guest.ckirduhx-cgu.1.rcgu.o)
            "___guest_request", referenced from:
                ___guest_call in libwapc_guest-3f6684dfb69c655b.rlib(wapc_guest-3f6684dfb69c655b.wapc_guest.ckirduhx-cgu.1.rcgu.o)
            "___guest_response", referenced from:
                ___guest_call in libwapc_guest-3f6684dfb69c655b.rlib(wapc_guest-3f6684dfb69c655b.wapc_guest.ckirduhx-cgu.1.rcgu.o)
          ld: symbol(s) not found for architecture x86_64
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

From stackoverflow, I found a command:
```
cargo rustc --release -- -C link-arg=-undefined -C link-arg=dynamic_lookup
```

which builds just fine. After some research it just looks like this requires undefined symbols to be looked up at build time and ignore missing symbols 🤔

**Comment by autodidaddict** [comment]
Something that has `__guest_error` is a webAssembly module. If it's being
compiled without wasm that's what's causing that. This feels like a target
mismatch problem.

**Comment by autodidaddict** [comment]
```
apability-providers/nats-keyvalue on  kvcache-nats is 📦 v0.1.0 via 🦀 v1.48.0 
❯ cargo build
   Compiling proc-macro2 v1.0.24
   Compiling unicode-xid v0.2.1
   Compiling syn v1.0.54
   Compiling libc v0.2.81
   Compiling serde_derive v1.0.118
   Compiling serde v1.0.118
   Compiling cfg-if v0.1.10
   Compiling autocfg v1.0.1
   Compiling log v0.4.11
   Compiling getrandom v0.1.15
   Compiling byteorder v1.3.4
   Compiling memchr v2.3.4
   Compiling once_cell v1.5.2
   Compiling version_check v0.9.2
   Compiling typenum v1.12.0
   Compiling cc v1.0.66
   Compiling lazy_static v1.4.0
   Compiling futures-core v0.3.8
   Compiling fastrand v1.4.0
   Compiling spin v0.5.2
   Compiling ryu v1.0.5
   Compiling cache-padded v1.1.1
   Compiling untrusted v0.7.1
   Compiling futures-io v0.3.8
   Compiling ppv-lite86 v0.2.10
   Compiling parking v2.0.0
   Compiling pin-project-lite v0.1.11
   Compiling waker-fn v1.1.0
   Compiling itoa v0.4.6
   Compiling serde_json v1.0.60
   Compiling signature v1.2.2
   Compiling event-listener v2.5.1
   Compiling subtle v2.3.0
   Compiling opaque-debug v0.3.0
   Compiling cpuid-bool v0.1.2
   Compiling vec-arena v1.0.0
   Compiling async-task v4.0.3
   Compiling unicode-xid v0.1.0
   Compiling cfg-if v1.0.0
   Compiling base64 v0.12.3
   Compiling atomic-waker v1.0.0
   Compiling regex-syntax v0.6.21
   Compiling openssl-probe v0.1.2
   Compiling data-encoding v2.3.1
   Compiling base64 v0.13.0
   Compiling termcolor v1.1.2
   Compiling json v0.12.4
   Compiling humantime v2.0.1
   Compiling thread_local v1.0.1
   Compiling wapc-guest v0.4.0
   Compiling num-traits v0.2.14
   Compiling num-integer v0.1.44
   Compiling concurrent-queue v1.2.2
   Compiling generic-array v0.14.4
   Compiling ed25519 v1.0.3
   Compiling async-lock v2.3.0
   Compiling proc-macro2 v0.2.3
   Compiling base64-url v1.4.8
   Compiling ring v0.16.19
   Compiling async-channel v1.5.1
   Compiling quote v0.4.2
   Compiling quote v1.0.7
   Compiling futures-lite v1.11.2
   Compiling aho-corasick v0.7.15
   Compiling nb-connect v1.0.2
   Compiling time v0.1.44
   Compiling atty v0.2.14
   Compiling syn v0.12.15
   Compiling rand_core v0.5.1
   Compiling blocking v1.0.2
   Compiling async-executor v1.4.0
   Compiling regex v1.4.2
   Compiling rand_chacha v0.2.2
   Compiling rmp v0.8.9
   Compiling digest v0.9.0
   Compiling block-buffer v0.9.0
   Compiling rand v0.7.3
   Compiling sha2 v0.9.2
   Compiling eventsourcing-derive v0.1.3
   Compiling nuid v0.2.2
   Compiling uuid v0.8.1
   Compiling sct v0.6.0
   Compiling webpki v0.21.4
   Compiling synstructure v0.12.4
   Compiling zeroize_derive v1.0.1
   Compiling zeroize v1.2.0
   Compiling curve25519-dalek v3.0.0
   Compiling subtle-encoding v0.5.1
   Compiling signatory v0.21.0
   Compiling serde_bytes v0.11.5
   Compiling ed25519-dalek v1.0.1
   Compiling rmp-serde v0.15.0
   Compiling rmp-serde v0.14.4
   Compiling chrono v0.4.19
   Compiling polling v2.0.2
   Compiling rustls v0.18.1
   Compiling nkeys v0.0.11
   Compiling env_logger v0.8.2
   Compiling async-io v1.3.1
   Compiling async-net v1.5.0
   Compiling actor-core v0.2.0 (https://github.com/wasmCloud/actor-interfaces?branch=main#9565c6ae)
   Compiling eventsourcing v0.1.5
   Compiling actor-keyvalue v0.2.0 (https://github.com/wasmCloud/actor-interfaces?branch=main#9565c6ae)
   Compiling wascc-codec v0.9.0
   Compiling async-rustls v0.1.2
   Compiling rustls-native-certs v0.4.0
   Compiling nats v0.8.6
   Compiling nats-keyvalue v0.1.0 (/home/kevin/Code/wascc/capability-providers/nats-keyvalue)
    Finished dev [unoptimized + debuginfo] target(s) in 44.67s
```

File: nats-keyvalue/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,545 @@
...
+
+        let default_replay_req_subject = DEFAULT_REPLAY_REQ_SUBJECT.to_string();
+        let replay_req_subject = values
+            .get(STATE_REPLAY_SUBJECT_KEY)
+            .unwrap_or(&default_replay_req_subject)
+            .to_string();
+
+        let cache = self.cache.clone();
+        let cache2 = self.cache.clone();
+        let cache3 = self.cache.clone();
```
[the remark this answers is not in the document]
Comment (autodidaddict) [reply nats-keyvalue/src/lib.rs:145]: I cannot describe how much I loathe the "clone a bunch of stuff I'm about to send down the closure rabbit hole" pattern. There's _got_ to be a better way.

**Review by brooksmtownsend — APPROVED** [review]
LGTM

**Pull request merged**